In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, r2_score

In [2]:
df = pd.read_csv("../data/processed/cleaned_student_data.csv")
df.head()

,Study_Hours_per_Week,Attendance_Rate,Past_Exam_Scores,Parental_Education_Level,Internet_Access_at_Home,Extracurricular_Activities,Final_Exam_Score
0,31,68.267841,86,0,1,1,63
1,16,78.222927,73,3,0,0,50
2,21,87.525096,74,3,1,0,55
3,27,92.076483,99,1,0,0,65
4,37,98.655517,63,2,0,1,70


In [3]:
X = df.drop("Final_Exam_Score", axis=1)
y = df["Final_Exam_Score"]

In [4]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

In [5]:
model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

In [6]:
y_pred = model.predict(X_test)

In [7]:
from sklearn.metrics import mean_absolute_error, r2_score

mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("MAE:", mae)
print("R2 Score:", r2)

MAE: 3.0989015443988177
R2 Score: 0.6582043783921546


In [8]:
feature_importance = pd.DataFrame({
    "Feature": X.columns,
    "Coefficient": model.coef_
})

feature_importance.sort_values(by="Coefficient", ascending=False)

,Feature,Coefficient
5,Extracurricular_Activities,0.871625
0,Study_Hours_per_Week,0.274073
2,Past_Exam_Scores,0.236035
1,Attendance_Rate,0.224989
3,Parental_Education_Level,0.215142
4,Internet_Access_at_Home,-0.916713


### Model Performance

The linear regression model achieved an MAE of 3.0989015443988177 and an R² score of 0.6582043783921546, indicating that the model explains a meaningful portion of variation in final exam scores.

### Feature Influence

Past exam scores showed the strongest influence on prediction, while attendance and study hours also contributed positively.

# MODEL v2 (Scaled)

### Improved Model with Feature Scaling

In [9]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)

In [10]:
X_test_scaled = scaler.transform(X_test)

In [11]:
scaled_model = LinearRegression()
scaled_model.fit(X_train_scaled, y_train)

LinearRegression()

In [12]:
y_pred_scaled = scaled_model.predict(X_test_scaled)

In [13]:
mae_scaled = mean_absolute_error(y_test, y_pred_scaled)
r2_scaled = r2_score(y_test, y_pred_scaled)

print("Scaled MAE:", mae_scaled)
print("Scaled R2:", r2_scaled)

Scaled MAE: 3.0989015443988173
Scaled R2: 0.6582043783921546


In [14]:
feature_importance = pd.DataFrame({
    "Feature": X.columns,
    "Coefficient": scaled_model.coef_
})

feature_importance.sort_values(by="Coefficient", ascending=False)

,Feature,Coefficient
2,Past_Exam_Scores,3.424760
1,Attendance_Rate,3.120951
0,Study_Hours_per_Week,2.425783
5,Extracurricular_Activities,0.435788
3,Parental_Education_Level,0.235940
4,Internet_Access_at_Home,-0.457210


After feature scaling, past exam scores emerged as the strongest predictor of final exam performance, followed by attendance rate and study hours. Scaling revealed a more realistic ranking of feature influence by placing all variables on comparable numerical ranges. However, the model evaluation metrics, MAE and R2 remain the same.

In [15]:
import pickle

with open("../models/linear_model_scaled.pkl", "wb") as f:
    pickle.dump(scaled_model, f)

In [16]:
with open("../models/scaler.pkl", "wb") as f:
    pickle.dump(scaler, f)